src/System.Management.Automation/engine/parser/DynamicKeyword.cs

```diff
@@ -0,0 +1,2759 @@
+    /// </summary>
+    public enum DynamicKeywordNameMode
+    {
+        /// <summary>
+        /// This keyword does not take a name value
+        /// </summary>
+        NoName = 0,
+        /// <summary>
+        /// Name must be present and simple non-empty bare word
+        /// </summary>
+        SimpleNameRequired = 1,
+        /// <summary>
+        /// Name must be present but can also be an expression
+        /// </summary>
+        NameRequired = 2,
+        /// <summary>
+        /// Name may be optionally present, but if it is present, it must be a non-empty bare word.
+        /// </summary>
+        SimpleOptionalName = 3,
+        /// <summary>
+        /// Name may be optionally present, expression or bare word
+        /// </summary>
+        OptionalName = 4,
+    };
+
+    /// <summary>
+    /// Defines the body mode for a dynamic keyword. It can be a scriptblock, hashtable or command which means no body
+    /// </summary>
+    public enum DynamicKeywordBodyMode
+    {
+        /// <summary>
+        /// The keyword act like a command
+        /// </summary>
+        Command = 0,
+        /// <summary>
+        /// The keyword has a scriptblock body
+        /// </summary>
+        ScriptBlock = 1,
+        /// <summary>
+        /// The keyword has hashtable body
+        /// </summary>
+        Hashtable = 2,
+    }
+
+    /// <summary>
+    /// Defines the use semantics of a dynamic keyword for a given block
+    /// </summary>
+    public enum DynamicKeywordUseMode
+    {
+        /// <summary>
+        /// The keyword must be used exactly once in a block
+        /// </summary>
+        Required = 0,
+
+        /// <summary>
+        /// The keyword must be used at least once in a block
+        /// </summary>
+        RequiredMany = 1,
+
+        /// <summary>
+        /// The keyword may be used 0 or 1 times in a block
+        /// </summary>
+        Optional = 2,
+
+        /// <summary>
+        /// The keyword may be used zero or more times in a block (i.e. there are no use restrictions)
+        /// </summary>
+        OptionalMany = 3,
+    }
+
+    /// <summary>
+    /// Defines the schema/behaviour for a dynamic keyword.
+    /// a constrained
+    /// </summary>
+    public class DynamicKeyword
+    {
+        #region static properties/functions
+
+        /// <summary>
+        /// Stack defining a cache of DynamicKeywordNamespaces. Note that this may behave strangely if
+        /// pushed or popped while in a scope -- this may need to be looked at (TODO)
+        /// </summary>
+        private static Stack<DynamicKeywordNamespace> DynamicKeywordNamespaceStack
+        {
+            get
+            {
+                return t_dynamicKeywordNamespaceStack ??
+                    (t_dynamicKeywordNamespaceStack = new Stack<DynamicKeywordNamespace>());
+            }
+        }
+        //[ThreadStatic]
```

Comment:
That was an oversight -- we keep `ThreadStatic` currently, mainly because I'm concerned that removing it will impact the static API in some unforeseen way.